### DISCUSSION TITLE
How to use --preload-file when compiling LLVM with emscripten?

### DISCUSSION THREAD

**Discussion by AlistairKeiller** [emscripten-core/emscripten#17579] [Q&A]
I am working on a browser-based c++ IDE/editor. To achieve this, I need to run a c++ compiler in the browser. 

I have gotten LLVM compiled to webassembly with emscripten. Now I need to provide LLVM with libc++. It seems like the cleanest way to achieve this is to preload https://github.com/WebAssembly/wasi-sdk/releases/download/wasi-sdk-16/wasi-sysroot-16.0.tar.gz in the MEMFS using emscripten's `--preload-file wsysroot` flag.

However, adding this flag to the LDFLAGS environment variable during the CMake step causes the error `CMake Error at cmake/modules/CheckCompilerVersion.cmake:98 (message):   libstdc++ version must be at least 5.1` ( https://github.com/AlistairKeiller/editor-wasm/tree/66b2a22e4865e6791252ff4923c78b7f24b101c8 ). This is the same error that occurs when I used the flag `s ALLOW_MEMORY_GROWTH` instead of `-s ALLOW_MEMORY_GROWTH`, so it seems to be caused by invalid flags.

I also tried installing libcxx and enabling the libcxx build target in CMake, which just gave the same error ( https://github.com/AlistairKeiller/editor-wasm/tree/3874fd76b3c3f77e50916aaca91eafe37ca25166 ).

Any suggestions on how to preload this file into the LLVM's MEMFS?

**Comment by sbc100** [comment]
How are you setting your `LDFLAGS`?  

I would recommend avoiding spaces in your link flags.  Prefer `-sALLOW_MEMORY_GROWTH` and `--preload-file=wsysroot`.

I would recommend `--embed-file` over `--preload-file` since the data will then be directly embedded in the wasm file.  It will make the wasm file big but the file access will be very fast.

**Reply by AlistairKeiller** [reply]
I am setting the `LDFLAGS` envorment variable. I have also tried setting the `CMAKE_CXX_FLAGS` CMake flag with similar results.

Thanks for the formatting tip, although that did not change the result.

Thanks for the suggestion to use `--embed-file`. Trying to use `--embed-file` also has the same error as using `--preload-file`.

To fix the libstdc++ version error from before, I added the flag `-DLLVM_ENABLE_LIBCXX=ON`. This resulted in the error `Host compiler appears to require libatomic, but cannot find it` ( https://github.com/AlistairKeiller/editor-wasm/blob/e8eb0872840628cf431a1b39c142f9cbc82377be/build.sh ). Setting `-DLLVM_ENABLE_THREADS=OFF` does not change anything. According to a StackOverflow post, this is an unnecessary error caused by LLVM's messy build system, but it still causes CMake to exit and prevents me from preceding. I could override this check by patching `CMake/modules/CheckAtomic.cmake:59` to always evaluate to true, but this seems very messy and, therefore, will probably introduce other issues. Any suggestions?

**Reply by sbc100** [reply]
I think you need add `--embed-file` link flags only for the one target you care about.  Setting it in LDFLAGS globally is not going to work since it will likely break the experiments/tests/configuration phase of cmake.. those little programs don't want to be linked with those flags.

**Reply by sbc100** [reply]
If you haven't gotten llvm to build yet, maybe follow the recipes used others who have done it in past:  See https://github.com/emscripten-core/emscripten/issues/6432 and https://github.com/jprendes/emception for example

**Reply by AlistairKeiller** [reply]
I think the Emception project is really interesting and impressive. However, its method of preloading files, wasm-package, seems unnecessarily complicated, difficult to implement, and less performant than the `--embed-file` you mentioned. So I really want to try to use Emscripten's `--embed-file` method.

I see the issues with trying to use `--embed-file` as a global flag, so I tried setting it for just clang by patching clang's CMakeLists.txt file with both `set_source_files_properties` and `set_target_properties`. I could not get any feedback that `set_source_files_properties(clang.cpp PROPERTIES COMPILE_FLAGS --preload-file=.)` was working, and it did not seem to be changing the output from the compiler. `set_target_properties(clang PROPERTIES COMPILE_FLAGS --preload-file=.)` seems to be working because setting it early in CMakeLists.txt file gives an error that the clang target does not exist. When it is set later in the file, it does not appear to change the output.

Do you have any suggestions for how to properly set the flags for the specific target?

**Reply by sbc100** [reply]
`--preload-file` is a linker flag rather than a compiler flag.  I imagine you would need to find some way add the link flag to just the clang binary.
